panther_manager/test/test_safety_behavior_tree.cpp

```diff
@@ -0,0 +1,422 @@
+#include "rclcpp/rclcpp.hpp"
+
+#include "sensor_msgs/msg/battery_state.hpp"
+#include "std_msgs/msg/bool.hpp"
+#include "std_srvs/srv/set_bool.hpp"
+#include "std_srvs/srv/trigger.hpp"
+
+#include "panther_msgs/msg/driver_state.hpp"
+#include "panther_msgs/msg/io_state.hpp"
+#include "panther_msgs/msg/system_status.hpp"
+
+#include <include/test_behavior_tree_utils.hpp>
+#include <panther_manager/safety_manager_node.hpp>
+
+using BoolMsg = std_msgs::msg::Bool;
+using BatteryStateMsg = sensor_msgs::msg::BatteryState;
+using IOStateMsg = panther_msgs::msg::IOState;
+using DriverStateMsg = panther_msgs::msg::DriverState;
+using SystemStatusMsg = panther_msgs::msg::SystemStatus;
+using SetBoolSrv = std_srvs::srv::SetBool;
+using TriggerSrv = std_srvs::srv::Trigger;
+
+class SafetyManagerNodeWrapper : public panther_manager::SafetyManagerNode
+{
+public:
+  SafetyManagerNodeWrapper(
+    const std::string & node_name, const rclcpp::NodeOptions options = rclcpp::NodeOptions())
+  : SafetyManagerNode(node_name, options)
+  {
+  }
+
+  ~SafetyManagerNodeWrapper() {}
+
+  BT::NodeStatus GetSafetyTreeStatus() { return this->safety_tree_status_; }
+  BT::NodeStatus GetShutdownTreeStatus() { return this->shutdown_tree_status_; }
+};
+
+class TestSafetyBehaviorTree : public testing::Test
+{
+public:
+  TestSafetyBehaviorTree();
+  ~TestSafetyBehaviorTree();
+
+protected:
+  void SpinWhileRunning();
+  void PublishDefaultStates();
+  void PublishEStop(const bool data);
+  void PublishBatteryState(
+    const std::uint8_t status, const std::uint8_t health, const float percentage);
+  void PublishIOState(const bool fan, const bool aux_power);
+  void PublishDriverState(const float temperature);
+  void PublishSystemStatus(const float cpu_temp);
+
+  static constexpr float kFanTurnOffTimeout = 1.0;
+
+  bool fan_state_ = false;
+  bool aux_power_state_ = true;
+  bool e_stop_triggered_ = false;
+  std::shared_ptr<SafetyManagerNodeWrapper> safety_manager_node_;
+
+private:
+  void FanServerCB(
+    const SetBoolSrv::Request::SharedPtr & request, SetBoolSrv::Response::SharedPtr response);
+  void AUXPowerServerCB(
+    const SetBoolSrv::Request::SharedPtr & request, SetBoolSrv::Response::SharedPtr response);
+  void EStopTriggerSrvCB(
+    const TriggerSrv::Request::SharedPtr & /*request*/, TriggerSrv::Response::SharedPtr response);
+
+  rclcpp::Publisher<BoolMsg>::SharedPtr e_stop_pub_;
+  rclcpp::Publisher<BatteryStateMsg>::SharedPtr battery_state_pub_;
+  rclcpp::Publisher<IOStateMsg>::SharedPtr io_state_pub_;
+  rclcpp::Publisher<DriverStateMsg>::SharedPtr driver_state_pub_;
+  rclcpp::Publisher<SystemStatusMsg>::SharedPtr system_status_pub_;
+
+  rclcpp::Service<SetBoolSrv>::SharedPtr fan_server_;
+  rclcpp::Service<SetBoolSrv>::SharedPtr aux_power_server_;
+  rclcpp::Service<TriggerSrv>::SharedPtr e_stop_trigger_server_;
+};
+
+TestSafetyBehaviorTree::TestSafetyBehaviorTree()
+{
+  using namespace std::placeholders;
+
+  rclcpp::init(0, nullptr);
```

Comment:
How about initializing the rclcpp once in the main function?

Reply:
The reason for this is that shutdown tree calls `rclcpp::shutdown()`. So to test shutdown behavior init has to be called for each test separately.